demos/GMW_demo/gmw_and.js

```diff
@@ -0,0 +1,175 @@
+
+/**
+ * AND on two shares, return secret share of ci.
+ * @function gmw_and
+ * @param {module:jiff-client~JIFFClient} jiff - the jiff instance
+ * @param {number} secret - the secret1 to share.
+ * @param {number} secret - the secret2 to share.
+ * @returns {object} a secret share of value ci
+ */
+
+function gmw_and(jiff,share1,share2) {
+  if (!(share1.jiff === share2.jiff)) {
+    throw new Error('shares do not belong to the same instance (^)');
+  }
+  if (!share1.jiff.helpers.Zp_equals(share1, share2)) {
+    throw new Error('shares must belong to the same field (^)');
+  }
+  if (!share1.jiff.helpers.array_equals(share1.holders, share2.holders)) {
+    throw new Error('shares must be held by the same parties (^)');
+  }
+
+  // AND the two shares when ready
+  var ready = function () {
+    var final_deferred = new jiff.helpers.Deferred();
+    var final_promise = final_deferred.promise;
+    var wi= share1.value & share2.value;
+    var csecret=[];
+    csecret.push(share1.value);
+    csecret.push(share2.value);
+    // receivinglist of OT msg
+    var recls=[];
+    var i;
+    for ( i=1;i<=jiff.party_count;i++) {
+      if (i<jiff.id) {
+        recls.push(i);
+      }
+    }
+
+    var share_id ;
+    if (share_id==null) {
+      share_id= jiff.counters.gen_op_id2('otshare', share1.holders, share1.holders);
+    }
+    var my_count;//list of promises needs to be resolve
+    if (typeof (my_count) === 'undefined') {
+      my_count= initial_deferreds(jiff,share_id);
+    }
+    // generate OT msg
+    var four_opts={};
+    four_opts['opts']=OTGate(csecret);// jason object
+    four_opts['sender_id']=jiff.id;
+    four_opts['op_id']=share_id;
+    four_opts = jiff.hooks.execute_array_hooks('beforeOperation', [jiff, 'open', four_opts], 2);
+    var mymsg=JSON.stringify(four_opts);
+    var help=[];
+    help.push(wi);
+    //wi resolve local wi; special case: jiff.id===jiff.party_count, return value
+    jiff.deferreds[share_id][jiff.id].resolve(help);
+
+    // OT send process, send OT msg to small party_ids
+    jiff.emit('OT',recls,mymsg,true);
```

Comment:
Let's re-organize this to rely on the 1-out-of-N OT library (https://github.com/wyatt-howe/1-out-of-n). We can also abstract some of the communication code to a separate file/function and make this function smaller.

i.e let's have an 'IO.js' file that implements get/give (a la https://github.com/wyatt-howe/1-out-of-n/blob/master/demo/io-template.js). This `emit`call can be placed in that `give()` function. at the top of this file we can require both the OT library as `var OT = require(...)l` and IO.js as `var IO = require('IO.js');`. 

then in this function (gmw_and) we can simply loop through the party_ids we want to send to and call OT.send for each one:
```JavaScript
for (var i = 1; i < jiff.party_id; i++) {
  //construct OPTS = [options 1-4]
  //
  OT.send(OPTS, 4, op_id); //op_id should be the op_id generated above + party we're sending to 
 // we have to add the party_id to the op_id here to work around how the OT library works ( to send to the correct person)
}
```
Within IO.js, get can call `jiff.emit('OT', ...)` similar to how you do here, you will need to parse the end of the op_id to figure out exactly who the message is going to (again, this is sort of a hack to get around how the OT library is designed)

Reply:
It seems currently that the emit is sent once to all parties, I think one unique message should be sent to each party individually

Reply:
even if the options (OPTS = [..., ..., ..., ...]) are the same for each message sent ( one to each party) it is important for security that an individual, pair-wise OT is performed with each receiving party - which is why we should rely on the library and make n-party_id calls to OT.send()